File: scripts/check-native-quit.py
#!/usr/bin/env python3
"""Check an already-built native AppKit smoke executable in a child process."""
import subprocess
import sys

result = subprocess.run(sys.argv[1:], capture_output=True, text=True, timeout=30)
sys.stdout.write(result.stdout)
sys.stderr.write(result.stderr)
expected = [
    "cancel-kept-pty-alive",
    "repeat-coalesced",
    "retry-kept-pty-alive",
    "capture-before-cleanup",
    "approved",
    "will-terminate-after-cleanup",
]
actual = [line.removeprefix("NATIVE_QUIT_SMOKE ") for line in result.stdout.splitlines()
          if line.startswith("NATIVE_QUIT_SMOKE ")]
if result.returncode != 0 or actual != expected:
    raise SystemExit(f"native quit smoke failed: exit={result.returncode}, markers={actual!r}")

File: scripts/check-native-quit.ts
import { checkSmokeProcess, runSmokeProcess } from "./smoke-process.ts";
/** Check an already-built native AppKit smoke executable in a child process. */
export function checkNativeQuit(exitCode: number, output: string): void {
  const expected = [
    "cancel-kept-pty-alive",
    "repeat-coalesced",
    "retry-kept-pty-alive",
    "capture-before-cleanup",
    "approved",
    "will-terminate-after-cleanup",
  ];
  const prefix = "NATIVE_QUIT_SMOKE ";
  const actual = output.split(/\r?\n/)
    .filter(line => line.startsWith(prefix))
    .map(line => line.slice(prefix.length));
  if (exitCode !== 0 || JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`native quit smoke failed: exit=${exitCode}, markers=${JSON.stringify(actual)}`);
  }
}

if (import.meta.main) {
  const command = Bun.argv.slice(2);
  if (command.length === 0) throw new Error("usage: check-native-quit.ts <executable> [args...]");
  const result = await runSmokeProcess(command, { timeoutMs: 30_000 });
  checkSmokeProcess(result, "native quit smoke");
  checkNativeQuit(result.exitCode, result.stdout);
}
